internal/provider/vm/internal_disk_resource.go

```diff
@@ -0,0 +1,618 @@
+// Package vm provides a Terraform resource.
+package vm
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/int64planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/vmware/go-vcloud-director/v2/govcd"
+	govcdtypes "github.com/vmware/go-vcloud-director/v2/types/v56"
+
+	"github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/client"
+	"github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/helpers/boolpm"
+	"github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/provider/common/vapp"
+	"github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/provider/common/vm"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ resource.Resource                = &internalDiskResource{}
+	_ resource.ResourceWithConfigure   = &internalDiskResource{}
+	_ resource.ResourceWithImportState = &internalDiskResource{}
+)
+
+// NewInternalDiskResource is a helper function to simplify the provider implementation.
+func NewInternalDiskResource() resource.Resource {
+	return &internalDiskResource{}
+}
+
+// internalDiskResource is the resource implementation.
+type internalDiskResource struct {
+	client *client.CloudAvenue
+}
+
+type internalDiskResourceModel struct {
+	ID              types.String `tfsdk:"id"`
+	VDC             types.String `tfsdk:"vdc"`
+	VAppName        types.String `tfsdk:"vapp_name"`
+	VMName          types.String `tfsdk:"vm_name"`
+	BusType         types.String `tfsdk:"bus_type"`
+	SizeInMb        types.Int64  `tfsdk:"size_in_mb"`
+	BusNumber       types.Int64  `tfsdk:"bus_number"`
+	UnitNumber      types.Int64  `tfsdk:"unit_number"`
+	ThinProvisioned types.Bool   `tfsdk:"thin_provisioned"`
+	Iops            types.Int64  `tfsdk:"iops"`
+	StorageProfile  types.String `tfsdk:"storage_profile"`
+	AllowVMReboot   types.Bool   `tfsdk:"allow_vm_reboot"`
+}
+
+// Metadata returns the resource type name.
+func (r *internalDiskResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_vm_internal_disk"
+}
+
+// Schema defines the schema for the resource.
+func (r *internalDiskResource) Schema(ctx context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
```

Comment:
This schema is required in `cloudavenue_vapp_vm`. It's possible to move Schema in the `func` ?